### PR TITLE
Move jni-fuse to a independent project

### LINKED ISSUE #12852: Make jni-fuse native to a new module to reduce the code
**Is your feature request related to a problem? Please describe.**

Now, we have a great module `integration/fuse`, it supply a POSIX interface through FUSE, further more, Alluxio invent a JNI-FUSE other than JNR-FUSE, and leverage it, Alluxio FUSE can be better than before.

But now, the JNI-FUSE bring the native code to alluxio repository, and many developer are not interested in JNI-FUSE native code, but they have to do some effort to prepare a build env.

**Describe the solution you'd like**

Create a single project or module, let's say it named `jni-fuse`, alluxio will deploy it into central maven repository, and supply a profile to let developer use the jni-fuse instead of build it.

**Describe alternatives you've considered**

Creaate a single project or module for JNI-FUSE.

**Urgency**

Normal

**Additional context**
NO.

### PR DESCRIPTION
**Pull request by maobaolong**
Fix #12852

This PR move jni-fuse related code to this project https://github.com/maobaolong/jnifuse (POC now, we can maintain it under alluxio github organization),

Meanwhile, jnifuse supply two module `jnifuse-java` and `jnifuse-native`, the output jar of jnifuse-native `jnifuse-native-1.0.0.jar` will contains native jni lib.

Leverage this PR, alluxio main repository will never responsible for the building of native code.

### CODE REVIEW

**Comment by maobaolong** [comment]
@apc999 Please take a look at this PR and related jni-fuse project. Thanks!

**Comment by maobaolong** [comment]
image

Why the native code cannot find the FuseFillDir class?

image

File: integration/jnifuse/jnifuse-java/src/main/java/alluxio/jnifuse/AbstractFuseFileSystem.java [hunk 1]
```diff
@@ -12,11 +12,11 @@
 package alluxio.jnifuse;
 
 import alluxio.jnifuse.struct.FileStat;
+import alluxio.jnifuse.struct.FuseContext;
 import alluxio.jnifuse.struct.FuseFileInfo;
 import alluxio.jnifuse.struct.Statvfs;
 import alluxio.jnifuse.utils.SecurityUtils;
-import alluxio.util.OSUtils;
```
Comment (apc999) [inline integration/jnifuse/jnifuse-java/src/main/java/alluxio/jnifuse/AbstractFuseFileSystem.java:18]: an empty line is required

File: integration/jnifuse/jnifuse-java/src/main/java/alluxio/jnifuse/constants/Errno.java [hunk 2]
```diff
@@ -105,7 +105,8 @@
   static final class StringTable {
     public static final java.util.Map<Errno, String> descriptions = generateTable();
     public static java.util.Map<Errno, String> generateTable() {
-      java.util.Map<Errno, String> map = new java.util.EnumMap<Errno, String>(Errno.class);
+      java.util.Map<Errno, String> map = new java.util.EnumMap<Errno, String>(
```
Comment (apc999) [inline integration/jnifuse/jnifuse-java/src/main/java/alluxio/jnifuse/constants/Errno.java:108]: revert the change?

File: integration/jnifuse/jnifuse-java/src/main/java/alluxio/jnifuse/LibFuse.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    if (libraryLoaded.get() == LibraryState.LOADED) {
+      return;
+    }
+
+    if (libraryLoaded.compareAndSet(LibraryState.NOT_LOADED,
+        LibraryState.LOADING)) {
+      String tmpDir = System.getenv("JNIFUSE_SHAREDLIB_DIR");
+      // loading possibly necessary libraries.
+      try {
+        System.loadLibrary("jnifuse");
```
Comment (apc999) [inline integration/jnifuse/jnifuse-java/src/main/java/alluxio/jnifuse/LibFuse.java:49]: shall we return if succeed ?

File: integration/jnifuse/jnifuse-java/src/main/java/alluxio/jnifuse/LibFuse.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+      } catch (UnsatisfiedLinkError e) {
+        // since it may be optional, we ignore its loading failure here.
+      }
+      try {
+        NativeLibraryLoader.getInstance().loadLibrary(tmpDir);
+      } catch (IOException e) {
```
Comment (apc999) [inline integration/jnifuse/jnifuse-java/src/main/java/alluxio/jnifuse/LibFuse.java:55]: shall we handle UnsatisfiedLinkError or other possible Throwables here?

File: integration/jnifuse/jnifuse-java/src/main/java/alluxio/jnifuse/utils/Environment.java [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+  public static boolean isPowerPC() {
+    return ARCH.contains("ppc");
+  }
+
+  public static boolean isS390x() {
+    return ARCH.contains("s390x");
+  }
+
+  public static boolean isWindows() {
+    return (OS.contains("win"));
```
Comment (apc999) [inline integration/jnifuse/jnifuse-java/src/main/java/alluxio/jnifuse/utils/Environment.java:31]: unnecessary outer `()`, same comments for below `returns`

File: integration/jnifuse/jnifuse-java/pom.xml [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    <version>2.5.0-SNAPSHOT</version>
+  </parent>
+
+  <artifactId>jnifuse-java</artifactId>
+
+  <properties>
+    <!-- The following paths need to be defined here as well as in the parent pom so that mvn can -->
+    <!-- run properly from sub-project directories -->
+    <build.path>${project.parent.parent.parent.basedir}/build</build.path>
+    <jnifuse-native.version>${project.version}</jnifuse-native.version>
```
Comment (apc999) [inline integration/jnifuse/jnifuse-java/pom.xml:27]: why do we need different version?

File: integration/jnifuse/jnifuse-native/pom.xml [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+-->
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+  <parent>
+    <artifactId>jnifuse</artifactId>
+    <groupId>org.alluxio</groupId>
+    <version>2.5.0-SNAPSHOT</version>
+  </parent>
+
+  <artifactId>jnifuse-native</artifactId>
```
Comment (apc999) [inline integration/jnifuse/jnifuse-native/pom.xml:21]: `integration-jnifuse-native`?

File: integration/jnifuse/pom.xml [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+    <profile>
+      <id>fuseNative</id>
+      <activation>
+        <activeByDefault>true</activeByDefault>
+      </activation>
+      <properties>
+        <runningWithNative>true</runningWithNative>
+      </properties>
+      <modules>
+        <module>jnifuse-java</module>
```
Comment (apc999) [inline integration/jnifuse/pom.xml:35]: unnecessary to include `jnifuse-java`, profile is additive

**Review by apc999 — CHANGES_REQUESTED** [review]
Thanks for the contribution!

File: integration/jnifuse/jnifuse-java/src/main/java/alluxio/jnifuse/LibFuse.java (hunk 4 above)
Comment (maobaolong) [reply integration/jnifuse/jnifuse-java/src/main/java/alluxio/jnifuse/LibFuse.java:55]: NativeLibraryLoader#loadLibrary has catch and handle `UnsatisfiedLinkError`, the handle logic could throw IOException, so we catch IOException here and mark as NOT_LOADED. The following is `NativeLibraryLoader#loadLibrary `

```java
  public synchronized void loadLibrary(final String tmpDir) throws IOException {
    try {
      System.loadLibrary(sharedLibraryName);
    } catch (final UnsatisfiedLinkError ule1) {
      try {
        System.loadLibrary(jniLibraryName);
      } catch (final UnsatisfiedLinkError ule2) {
        loadLibraryFromJar(tmpDir);
      }
    }
  }
```

File: integration/jnifuse/jnifuse-java/pom.xml (hunk 6 above)
Comment (maobaolong) [reply integration/jnifuse/jnifuse-java/pom.xml:27]: Maybe we can use jnifuse-native from maven repository, but build jnifuse-java in the alluxio project.

File: integration/jnifuse/jnifuse-native/pom.xml (hunk 7 above)
Comment (maobaolong) [reply integration/jnifuse/jnifuse-native/pom.xml:21]: -> alluxio-integration-jnifuse-native

File: integration/jnifuse/pom.xml (hunk 8 above)
Comment (maobaolong) [reply integration/jnifuse/pom.xml:35]: Yeah, exactly!

File: integration/jnifuse/jnifuse-java/src/main/java/alluxio/jnifuse/utils/Environment.java (hunk 5 above)
Comment (maobaolong) [reply integration/jnifuse/jnifuse-java/src/main/java/alluxio/jnifuse/utils/Environment.java:31]: Sure.

File: integration/fuse/pom.xml [hunk 9]
```diff
@@ -53,6 +53,11 @@
       <artifactId>alluxio-core-common</artifactId>
       <version>${project.version}</version>
     </dependency>
+    <dependency>
+      <groupId>org.alluxio</groupId>
+      <artifactId>jnifuse-java</artifactId>
```
[the remark this answers is not in the document]
Comment (maobaolong) [reply integration/fuse/pom.xml:58]: Done.

File: integration/jnifuse/jnifuse-java/pom.xml [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+-->
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+  <parent>
+    <artifactId>jnifuse</artifactId>
+    <groupId>org.alluxio</groupId>
+    <version>2.5.0-SNAPSHOT</version>
+  </parent>
+
+  <artifactId>jnifuse-java</artifactId>
```
[the remark this answers is not in the document]
Comment (maobaolong) [reply integration/jnifuse/jnifuse-java/pom.xml:21]: Done.

File: integration/jnifuse/jnifuse-java/pom.xml [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+    <dependency>
+      <groupId>org.slf4j</groupId>
+      <artifactId>slf4j-api</artifactId>
+    </dependency>
+    <dependency>
+      <groupId>org.apache.logging.log4j</groupId>
+      <artifactId>log4j-slf4j-impl</artifactId>
+    </dependency>
+  </dependencies>
+</project>
```
[the remark this answers is not in the document]
Comment (maobaolong) [reply integration/jnifuse/jnifuse-java/pom.xml:54]: Done.

File: integration/jnifuse/jnifuse-native/pom.xml [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+                  </fileset>
+                </copy>
+              </target>
+            </configuration>
+          </execution>
+        </executions>
+      </plugin>
+    </plugins>
+  </build>
+</project>
```
[the remark this answers is not in the document]
Comment (maobaolong) [reply integration/jnifuse/jnifuse-native/pom.xml:59]: Done.

File: integration/jnifuse/pom.xml [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+        <module>jnifuse-native</module>
+      </modules>
+    </profile>
+  </profiles>
+  <properties>
+    <!-- The following paths need to be defined here as well as in the parent pom so that mvn can -->
+    <!-- run properly from sub-project directories -->
+    <build.path>${project.parent.parent.basedir}/build</build.path>
+  </properties>
+</project>
```
[the remark this answers is not in the document]
Comment (maobaolong) [reply integration/jnifuse/pom.xml:45]: Done.

**Comment by maobaolong** [comment]
@apc999 Thanks for your review and suggestion, I've addressed your comments, PTAL.

Also, considering I bring jnifuse-native here, would you please help me how to build jnifuse-native through CI checks?

File: integration/jnifuse/jnifuse-java/src/main/java/alluxio/jnifuse/LibFuse.java (hunk 3 above)
Comment (maobaolong) [reply integration/jnifuse/jnifuse-java/src/main/java/alluxio/jnifuse/LibFuse.java:49]: These logic has been included in `NativeLibraryLoader.getInstance().loadLibrary(tmpDir);` so I will remove it.

**Comment by gpang** [comment]
What do you need to do to build this? You may need to update the github configs for building for github actions: https://github.com/Alluxio/alluxio/blob/master/.github/workflows/java8_integration_tests.yml

**Comment by maobaolong** [comment]
@gpang Thanks for your link, I really new for the github action, so I change my direction. I make it don't build and dependent native code by default now, so CI checks can passed.

Maybe we have to do some effort to create new dockerfile for build and deploy native code

**Comment by maobaolong** [comment]
@apc999 @gpang Thanks for the review before, today, I join the Alluxio fuse sync meeting and discuss this PR with @apc999 online, we think we can manual upload and maintain jni share libs for now, so that we can enjoy this feature in the next edge version.

Now, I do the following test.

- Run ` mvn clean install  -PfuseNative  -DskipTests -Dmaven.javadoc.skip=true -Dlicense.skip=true -Dcheckstyle.skip=true -Dfindbugs.skip=true`, and make sure `libjnifuse.dylib` can be regenerated.
- Run ` mvn clean install  -DskipTests -Dmaven.javadoc.skip=true -Dlicense.skip=true -Dcheckstyle.skip=true -Dfindbugs.skip=true`, and make sure haven't build `libjnifuse.dylib`
- Run the fat jar on macosx and linux, can mount, listing, read and write as expected.

File: integration/jnifuse/jnifuse-java/src/main/java/alluxio/jnifuse/utils/NativeLibraryLoader.java [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+import java.nio.file.Files;
+import java.nio.file.StandardCopyOption;
+
+/**
+ * This class is used to load the shared library from within the jar.
+ * The shared library is extracted to a temp folder and loaded from there.
+ */
+public class NativeLibraryLoader {
+  //singleton
+  private static final NativeLibraryLoader instance = new NativeLibraryLoader();
```
Comment (apc999) [inline integration/jnifuse/jnifuse-java/src/main/java/alluxio/jnifuse/utils/NativeLibraryLoader.java:26]:
```suggestion
  private static final NativeLibraryLoader INSTANCE = new NativeLibraryLoader();
```

File: integration/jnifuse/jnifuse-java/src/main/java/alluxio/jnifuse/utils/NativeLibraryLoader.java [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+  private static boolean initialized = false;
```
Comment (apc999) [inline integration/jnifuse/jnifuse-java/src/main/java/alluxio/jnifuse/utils/NativeLibraryLoader.java:27]: maybe a atomic boolean?

File: integration/jnifuse/jnifuse-java/src/main/java/alluxio/jnifuse/utils/NativeLibraryLoader.java [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+
+  private static final String sharedLibraryName =
```
Comment (apc999) [inline integration/jnifuse/jnifuse-java/src/main/java/alluxio/jnifuse/utils/NativeLibraryLoader.java:29]: same here, please follow Alluxio var naming convention here for static final member vars

File: integration/jnifuse/jnifuse-java/src/main/java/alluxio/jnifuse/utils/Environment.java [hunk 17]
```diff
@@ -0,0 +1,124 @@
...
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.jnifuse.utils;
+
+public class Environment {
```
Comment (apc999) [inline integration/jnifuse/jnifuse-java/src/main/java/alluxio/jnifuse/utils/Environment.java:14]: Alluxio code convention:
```
public final class Environment {
  private Environment() {} // Prevent instantiation
```

File: integration/jnifuse/jnifuse-java/src/main/java/alluxio/jnifuse/utils/Environment.java [hunk 18]
```diff
@@ -0,0 +1,124 @@
...
+  private static String OS = System.getProperty("os.name").toLowerCase();
```
Comment (apc999) [inline integration/jnifuse/jnifuse-java/src/main/java/alluxio/jnifuse/utils/Environment.java:15]: make OS and ARCH `final`?

File: integration/jnifuse/jnifuse-java/src/main/java/alluxio/jnifuse/utils/Environment.java (hunk 17 above)
Comment (apc999) [reply integration/jnifuse/jnifuse-java/src/main/java/alluxio/jnifuse/utils/Environment.java:14]: also, if this file is mostly derived from rocksdb lib, please acknowledge the original file in class comments

File: integration/jnifuse/jnifuse-java/src/main/java/alluxio/jnifuse/LibFuse.java [hunk 19]
```diff
@@ -0,0 +1,67 @@
...
+public class LibFuse {
+
+  private enum LibraryState {
+    NOT_LOADED,
+    LOADING,
+    LOADED
+  }
+
+  private static AtomicReference<LibraryState> libraryLoaded
+      = new AtomicReference<>(LibraryState.NOT_LOADED);
```
Comment (apc999) [inline integration/jnifuse/jnifuse-java/src/main/java/alluxio/jnifuse/LibFuse.java:29]:
```suggestion
  private static AtomicReference<LibraryState> libraryLoaded =
        new AtomicReference<>(LibraryState.NOT_LOADED);
```

File: integration/jnifuse/pom.xml [hunk 20]
```diff
@@ -0,0 +1,31 @@
...
+  <modelVersion>4.0.0</modelVersion>
+  <parent>
+    <groupId>org.alluxio</groupId>
+    <artifactId>alluxio-integration</artifactId>
+    <version>2.5.0-SNAPSHOT</version>
+  </parent>
+  <packaging>pom</packaging>
+  <artifactId>alluxio-integration-jnifuse</artifactId>
+  <modules>
+    <module>jnifuse-java</module>
```
Comment (apc999) [inline integration/jnifuse/pom.xml:23]: I would name these two modules

```
<module>fs</module>
<module>native</module>
```

jnif is already showing in their parent module name, redundant here

File: integration/jnifuse/jnifuse-native/pom.xml [hunk 21]
```diff
@@ -0,0 +1,68 @@
...
+                  <goal>run</goal>
+                </goals>
+                <configuration>
+                  <!--TODO:(maobaolong) make it can build for multi target-->
+                  <target>
+                    <mkdir dir="${project.build.directory}/native"/>
+                    <exec executable="make" failonerror="true" dir="${project.build.directory}/native">
+                      <arg line="-f ${project.basedir}/src/main/native/libjnifuse/Makefile"/>
+                    </exec>
+                    <copy todir="${project.build.directory}/classes">
```
Comment (apc999) [inline integration/jnifuse/jnifuse-native/pom.xml:54]: if the dynamic link libs are already in `alluxio-integration-jnifuse-native`, we can remove these two files in `lib/` and remove the library path set in https://github.com/Alluxio/alluxio/pull/12857 ?

File: integration/jnifuse/jnifuse-java/src/main/java/alluxio/jnifuse/AbstractFuseFileSystem.java [hunk 22]
```diff
@@ -34,11 +34,13 @@
  */
 public abstract class AbstractFuseFileSystem implements FuseFileSystem {
 
+  private static final Logger LOG = LoggerFactory.getLogger(
```
Comment (apc999) [inline integration/jnifuse/jnifuse-java/src/main/java/alluxio/jnifuse/AbstractFuseFileSystem.java:37]: unneccessary line wrap. Are you using line width of 80? Alluxio code convention is 100

**Review by apc999 — CHANGES_REQUESTED** [review]
more comments. thanks!

File: integration/jnifuse/jnifuse-java/src/main/java/alluxio/jnifuse/AbstractFuseFileSystem.java (hunk 22 above)
Comment (maobaolong) [reply integration/jnifuse/jnifuse-java/src/main/java/alluxio/jnifuse/AbstractFuseFileSystem.java:37]: Done

File: integration/jnifuse/jnifuse-native/pom.xml (hunk 21 above)
Comment (maobaolong) [reply integration/jnifuse/jnifuse-native/pom.xml:54]: Sure

File: integration/jnifuse/pom.xml (hunk 20 above)
Comment (maobaolong) [reply integration/jnifuse/pom.xml:23]: Done

File: integration/jnifuse/jnifuse-java/src/main/java/alluxio/jnifuse/utils/Environment.java (hunk 18 above)
Comment (maobaolong) [reply integration/jnifuse/jnifuse-java/src/main/java/alluxio/jnifuse/utils/Environment.java:15]: Done

File: integration/jnifuse/jnifuse-java/src/main/java/alluxio/jnifuse/utils/Environment.java (hunk 17 above)
Comment (maobaolong) [reply integration/jnifuse/jnifuse-java/src/main/java/alluxio/jnifuse/utils/Environment.java:14]: I add a comment like `This file mostly derived from rocksdb lib.`, so we don't need to rename it to `EnvironmentUtils `?

AFAIK, there are another file named `alluxio.util.EnvironmentUtils` already.

File: integration/jnifuse/jnifuse-java/src/main/java/alluxio/jnifuse/utils/NativeLibraryLoader.java (hunk 16 above)
Comment (maobaolong) [reply integration/jnifuse/jnifuse-java/src/main/java/alluxio/jnifuse/utils/NativeLibraryLoader.java:29]: done

File: integration/jnifuse/jnifuse-java/src/main/java/alluxio/jnifuse/utils/NativeLibraryLoader.java (hunk 15 above)
Comment (maobaolong) [reply integration/jnifuse/jnifuse-java/src/main/java/alluxio/jnifuse/utils/NativeLibraryLoader.java:27]: Done

**Comment by maobaolong** [comment]
@apc999 Thank you for the kindly review and lots of comments. I push a new commit to fix the style issue and the suggestion  rename refactor, PTAL.

**Review by apc999 — APPROVED** [review]

**Comment by apc999** [comment]
Thanks for the great contribution!

**Comment by apc999** [comment]
alluxio-bot, merge this please

**Comment by LuQQiu** [comment]
Thanks for making jni-fuse a independent project! @maobaolong 
Maybe I use the new updated code wrong, in the latest master branch,
I am not able to launch the jni fuse application in mac, are there new requirements for using this new module?

**Comment by maobaolong** [comment]
@LuQQiu Thanks for your information, Will check at once.

**Comment by maobaolong** [comment]
@LuQQiu I think I have passed my check locally.

 mvn clean install  -DskipTests -Dmaven.javadoc.skip=true -Dlicense.skip=true -Dcheckstyle.skip=true -Dfindbugs.skip=true -Denforcer.skip=true   -Djsse.enableSNIExtension=false

image

image

image

image

**Comment by LuQQiu** [comment]
My mac version is 4.0.5
Screen Shot 2021-02-25 at 6 39 53 PM

I ran 
```
$ mvn clean install -DskipTests -Dmaven.javadoc.skip=true -Dlicense.skip=true -Dcheckstyle.skip=true -Dfindbugs.skip=true -Denforcer.skip=true -Djsse.enableSNIExtension=false
$ integration/fuse/bin/alluxio-fuse mount fuseMount /
$ jps
24398 AlluxioMaster 
# couldn't find AlluxioFuse process
$ mount
java@macfuse0 on /Users/alluxio/alluxioFolder/alluxio/fuseMount (macfuse, nodev, nosuid, synchronous, mounted by alluxio)
$ ls /Users/alluxio/alluxioFolder/alluxio/fuseMount
ls: fuseMount: Device not configured
$ cd fuseMount && ls
.
$ cd /Users/alluxio/alluxioFolder/alluxio && bin/alluxio fs ls /
-rw-r--r--  alluxio        staff                    27040       PERSISTED 01-07-2021 10:27:40:410   0% /LICENSE2
```

**Comment by LuQQiu** [comment]
When i change the macOS version to 3.10.4, everything works fine! Thanks @maobaolong 
Any requirement for the Linux environment? which libfuse version does we need to install in the Linux env?
Does our current generate-tarball scripts support building the .so file for Linux environment?

**Comment by LuQQiu** [comment]
I am testing on linux env with fusermount version: 2.9.2
```
[ec2-user@AlluxioSandboxMaster-masters-1 alluxio]$ integration/fuse/bin/alluxio-fuse mount fuseMount /
Cannot find /tmp/alluxio/integration/fuse/bin/../target/alluxio-integration-fuse-2.5.0-SNAPSHOT-jar-with-dependencies.jar. Please compile alluxio with fuse profile and Java 8
```
The tarball was built with `./dev/scripts/generate-tarballs release `

Looks like fuse profile is needed?

**Pull request merged**